transformers4rec/torch/model/prediction_task.py

```diff
@@ -218,16 +218,22 @@ def forward(self, inputs: torch.Tensor, **kwargs):
             x = self.task_block(x)  # type: ignore
 
         # Retrieve labels either from masking or input module
-        if self.masking:
+        if self.masking and not ignore_masking:
             labels = self.masking.masked_targets
+            trg_flat = labels.flatten()
+            non_pad_mask = trg_flat != self.padding_idx
+            labels_all = torch.masked_select(trg_flat, non_pad_mask)
+            # remove padded items, keep only masked positions
+            x = self.remove_pad_3d(x, non_pad_mask)
         else:
+            # keep only last non-padded position for the 'Prediction step'
             labels = self.embeddings.item_seq
-
-        # remove padded items
-        trg_flat = labels.flatten()
-        non_pad_mask = trg_flat != self.padding_idx
-        labels_all = torch.masked_select(trg_flat, non_pad_mask)
-        x = self.remove_pad_3d(x, non_pad_mask)
+            non_pad_mask = labels != self.padding_idx
+            last_item_sessions = non_pad_mask.sum(dim=1) - 1
+            rows_ids = torch.arange(labels.size(0), dtype=torch.long, device=labels.device)
```

Comment:
Doesn't this change the behavior compared to what we had before? I understand that before when `self.masking == False` the full sequence of items were used as label. Was `self.masking == False` only for predict? Or do we have other use cases where masking should be False to take into account?

For `model.predict()` I think it makes sense to mask only the last item, assuming it is a dummy item added in the end of the sequence when using MLM just to enforce predicting the last item using the whole sequence.

Reply:
From our discussion yesterday, we should always set masking (`self.masking != None`) for the NextItemPredictionTask as the current implementation does not support explicit targets.  So I added a required check in the build method to make sure masking is present in the input Block. 

For the predict() step, we set the flag `ignore_masking=True` so that the whole sequence is considered and there is no need to add the dummy item in the end. The final prediction is computed based on the hidden states of the last position in the sequence (as the weights are shared between positions)